### PR TITLE
add data migration

### PR DESCRIPTION
**Pull request by sarmatdev**
- Add `migrateLegacyData`
- Wrote test

### CODE REVIEW

**Review by upacyxou — APPROVED** [review]
lgtm

File: contracts/OBSSStorage.sol [hunk 1]
```diff
@@ -371,6 +394,74 @@ contract OBSSStorage is Initializable, Context, ERC2771Recipient {
...
+        ++i;
+      }
+    }
+  }
+
+  function _addFeedLegacyReactionsBatch(
+    LegacyReaction[] memory legacyReactions
+  ) private {
+    uint256 length = legacyReactions.length;
+    for (uint8 i = 0; i < length; ) {
```
Comment (MixailE) [inline contracts/OBSSStorage.sol:431]: Looks like it's missing an incrementer `i++`

File: scripts/deploy.ts [hunk 2]
```diff
@@ -110,6 +178,90 @@ async function main() {
   )
 }
 
+const legacyContractAddress = '0x9e7A15E77e5E4f536b8215aaF778e786005D0f8d'
```
Comment (T-Damer) [inline scripts/deploy.ts:181]: I think it should be defined with a `.env`, so you more likely don't forget to update it

File: scripts/deploy.ts [hunk 3]
```diff
@@ -110,6 +178,90 @@ async function main() {
...
+    .attach(legacyContractAddress)
+    .connect(provider)
+    .connect(signer)
+
+  console.log(legacyContract.address)
+  const totalFeeds = await legacyContract.lastFeedId()
+  console.log(`Total feeds count: ${totalFeeds.toNumber()}`)
+  const legacyPosts: OBSSStorage.LegacyPostStruct[] = []
+  const legacyReactions: OBSSStorage.LegacyReactionStruct[] = []
+  for (let i = 0; i < 10; i++) {
```
Comment (T-Damer) [inline scripts/deploy.ts:198]: Why `i < 10`? What is `10`?

File: scripts/deploy.ts [hunk 4]
```diff
@@ -110,6 +178,90 @@ async function main() {
...
+    feedPosts.forEach(async (post: OBSSStorage.PostStructOutput, j) => {
+      legacyPosts.push({
+        post: {
+          author: post.author,
+          metadata: {
+            digest: post.metadata.digest,
+            hashFunction: post.metadata.hashFunction,
+            size: post.metadata.size,
+          },
+          commentsFeedId: 0,
```
Comment (T-Damer) [inline scripts/deploy.ts:217]: Shouldn't it use `post.feedId` or `post.commentsFeedId`?

File: scripts/deploy.ts [hunk 5]
```diff
@@ -110,6 +178,90 @@ async function main() {
...
+
+    const feedPosts = await legacyContract.getFeedPosts(
+      i,
+      0,
+      postsInFeed.toNumber()
+    )
+    console.log(`Feed ID: ${i}`)
+    feedPosts.forEach(async (post: OBSSStorage.PostStructOutput, j) => {
+      legacyPosts.push({
+        post: {
```
Comment (T-Damer) [inline scripts/deploy.ts:210]: Maybe we can do `post: { ...post, timestamp: post.timestamp.toNumber(), commentsFeedId: 0}`

File: scripts/deploy.ts [hunk 6]
```diff
@@ -110,6 +178,90 @@ async function main() {
...
+                hashFunction: post.metadata.hashFunction,
+                size: post.metadata.size,
+              },
+              author: '0x0000000000000000000000000000000000000000',
+              commentsFeedId: 0,
+              timestamp: 0,
+            },
+          })
+        }
+      } catch (_) {
```
Comment (T-Damer) [inline scripts/deploy.ts:253]: `catch (e) { console.error (e) }` is better, it will give you some info on what failed

File: test/utils.ts [hunk 7]
```diff
@@ -116,6 +117,29 @@ export function getFeedPostsBatch(length = 10) {
   return posts
 }
 
+export function getLegacyFeedPostsBatch(length = 10) {
+  const posts: OBSSStorage.LegacyPostStruct[] = []
+
+  for (let i = 0; i < length; i++) {
+    posts.push({
+      post: {
+        author: `0x000000000000000000000000000000000000000${i}`,
```
Comment (T-Damer) [inline test/utils.ts:126]: If we really need this `0x0000000000000000000000000000000000000000`, we should extract it into separate file and re-use

File: scripts/deploy.ts (hunk 3 above)
Comment (sarmatdev) [reply scripts/deploy.ts:198]: Test value, it's going to be `totalFeeds`

File: scripts/deploy.ts (hunk 4 above)
Comment (sarmatdev) [reply scripts/deploy.ts:217]: `commentsFeedId` hardcoded to zero to pass input validations. It's going to generate a new `commentsFeedId` on adding a feed in contract, like this
`uint256 commentsFeedId = addFeed(legacyPost.post.metadata);`

**Pull request merged**